### ISSUE TITLE
DNS Editor: TXT recods longer than 255 characters produce syntax error in bind

### ISSUE THREAD

**Issue by crazy4chrissi** [Froxlor/Froxlor#472]
# Summary

I created a TXT record in Froxlors DNS Editor for DKIM. The 2048bit key is longer than 255 characters.
The zone-file that the DNS Editor creates causes a syntax error and thus bind will not serve the zone anymore.

Long TXT records need to be split into multiple lines like this:

    a9d04665528b593d263a6e5256648c99._domainkey IN  1800 TXT (
         "k=rsa;p=MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAz2/ZfhxSI/A"
         "bqgh0amM8ylrlosirWeKShUhq7fg12aYmRwOqq9hIzO0Fcz1BzfgHVu6HU++rC5"
         "QoUK0JQK/nk4jwkDgvG2di2ZYmAvEbY/VeiK1x/TG0p1Iczr2k6Bj0gEAb/YGD2"
         "YbwrwAi4bDXwoPsYuuNn9TB3jjyWKu/dvOsqhff1/4Wc+FkOi0ClvgrXiklN28X"
         "TLjyjSyU794ntIoegXxrfwcwkhfPMvuqcnhfIC0Z8L71M4WR4SoHyNHVfBtNlUv"
         "VNROiXlMxtxnNQvfViSwz6LC8bYIxeAba3hSXPTChKu3qZtfR0o3jFwEWAfLQdg"
         "Ixler0jMEoAyJmfQIDAQAB")

[Source of this example](https://serverfault.com/questions/571720/publishing-long-domain-key-records-in-bind9)

# System information
* Froxlor version: 0.9.38.7-1+wheezy1
* Web server: apache2
* DNS server: Bind
* POP/IMAP server: Courier
* SMTP server: postfix
* FTP server: proftpd
* OS/Version: Debian Wheezy

# Steps to reproduce

1. Enable Nameserver for some domain
2. Open the DNS Editor
3. Create a new TXT record with more than 255 characters

# Expected behavior

1. A zone file with correct syntax should be created
2. Bind should reload without problem
3. The zone should still be served by bind

# Actual behavior

1. A zone file with syntax error is created
2. Bind reloads, detects syntax error
3. Bind stops serving the zone

# Log files/log entries
syslog:
<pre>
Aug 25 11:02:31 host3 named[17684]: dns_rdata_fromtext: /etc/bind/domains/myzone.de.zone:5: syntax error
Aug 25 11:02:31 host3 named[17684]: zone myzone.de/IN: loading from master file /etc/bind/domains/myzone.de.zone failed: syntax error
Aug 25 11:02:31 host3 named[17684]: zone myzone.de/IN: not loaded due to errors.
Aug 25 11:02:31 host3 named[17684]: reloading zones succeeded</pre>

# Proposed solution

1. Froxlor should split long records as explained above
2. Froxlor could check the generated zone file with `named-checkzone` to make sure no faulty zone gets loaded

**Comment by d00p** [comment]
Hey, thanks for the great report, appreciate it.

Could you test whether the following patch resolves the issue? -> https://gist.github.com/d00p/f2d97615aaea8f34e3fe4f1c6c8fe1c9

**Comment by crazy4chrissi** [comment]
Thanks for the fast patch, I will check it and tell you if it solves my problem.

However, from what I see, the patch only splits TXT records. If someone for some reason enters 300 characters in a SRV record, where I guess it could occur as well, he will get a syntax error as well. I have found this for SRV records:
https://help.blacknight.com/hc/en-us/articles/212512389-Creating-large-SPF-records-with-many-entries
So there is a workaround for SRV records. But Froxlor should not produce a zone file that produces a syntax error. I am not sure if SRV records can be split the same way as TXT, though.

**Comment by d00p** [comment]
I only know the multi-line stuff from TXT records

**Comment by crazy4chrissi** [comment]
I tried your patch. The first line start with two quotes and ends with two quotes. 
Bind complains like this:

    zone myzone.de/IN: loading from master file /etc/bind/domains/myzone.de.zone failed: unbalanced quotes
    zone myzone.de/IN: not loaded due to errors.


I guess `$this->content` already includes quotes. Either do not add another quote in the first and last line, or trim the quotes away at the beginning. Will make a slight difference in the length calculation, but I think it doesn't matter. I removed the additional quotes and then everything went fine.

Okay, I don't really care about other long record types, we can leave it for TXT only.

**Comment by crazy4chrissi** [comment]
I now also tried if the DKIM key really works. Sending a mail signed with this key to gmail worked. Google says the mail passed DKIM check, and my Thunderbird "DKIM Verifier" extension also says the DKIM signature is right. :-)

**Comment by d00p** [comment]
> I tried your patch. The first line start with two quotes and ends with two quotes.
I'm unable to reproduce that, even when entering the content with quotes in the dns-editor

**Comment by crazy4chrissi** [comment]
Maybe because you are trying with git version and I am trying with (patched) 0.9.38.7-1+wheezy1 ? I am currently checking where the quotes are coming from. My TXT records already have quotes in the `domain_dns_entries` table in the database. Yours don't?

**Comment by d00p** [comment]
No, no quotes in the table

**Comment by crazy4chrissi** [comment]
I found the `encloseTXTContent()` adds the quotes. The dns_editor.php calls it:
`$content = encloseTXTContent($content);` (line 142 for me)
And then this is:

```
function encloseTXTContent($txt_content, $isMultiLine = false)
{
	// check that TXT content is enclosed in " "
	if ($isMultiLine == false && Settings::Get('system.dns_server') != 'pdns') {
		if (substr($txt_content, 0, 1) != '"') {
			$txt_content = '"' . $txt_content;
		}
		if (substr($txt_content, - 1) != '"') {
			$txt_content .= '"';
		}
	}
	if (Settings::Get('system.dns_server') == 'pdns') {
		// no quotation for PowerDNS
		if (substr($txt_content, 0, 1) == '"') {
			$txt_content = substr($txt_content, 1);
		}
		if (substr($txt_content, - 1) == '"') {
			$txt_content = substr($txt_content, 0, -1);
		}
	}
	return $txt_content;
}
```
You are on powerDNS?

**Comment by d00p** [comment]
Well, I'm testing on demo.froxlor.org - and I'm just viewing the preview of the zone, which is exactly the same as froxlor would generate for BIND

**Comment by d00p** [comment]
Ok, got that little bugger :P This should do: https://gist.github.com/d00p/2c372256c1caee468e7dc502a5d44fca

**Comment by crazy4chrissi** [comment]
The preview of the zone also has two quotes for me. This is because `dns_editor.php` adds the quotes before storing the record in the database, depending on what server you use. It should not be done there, it should be done when generating the zone file. If you are on pdns, it will not add quotes, will not store them in the db and will not show them in the preview.
For me, it adds quotes in the DB, thus I get double quotes in the preview and zone file.

**Comment by crazy4chrissi** [comment]
Well, this fix solves the problem. But there is another problem I guess: If you switch from bind to pdns, your database contains quotes, and it will probably generate zone files that pdns does not like. Or if you switch from pdns to bind, it will generate bind zone files without quotes.
The database should be independent of the server that is used. The quotes should be added when generating the zone file, not when storing in the DB.

**Comment by d00p** [comment]
This is indeed correct....usually, you would not change your nameserver on production environments...but yes, the database should be clean. This will possibly a bit more work (updater needs to clean the 'false' records, etc.)

**Comment by crazy4chrissi** [comment]
Yes true. I don't care myself, and it is a different issue. I just wanted to mention it as we stumbled upon it.
